core/src/Kokkos_ExecPolicy.hpp

```diff
@@ -847,6 +847,301 @@ KOKKOS_INLINE_FUNCTION_DELETED Impl::TeamThreadRangeBoundariesStruct<
+  static constexpr MDTeamRangeParVector par_vector = ParVector;
+};
+
+// Tag class to keep track of the loop nest level and where to deploy thread and
+// vector parallelism
+//   - Rank is Kokkos::Rank<TotalRank, Iter>
+//     - TotalRank is the total number of loop nests
+//     - Iter is whether to go forward or backward through ranks (i.e. the
+//       iteration order for MDRangePolicy)
+//   - ParThreadRank is the nesting level on which to deploy thread parallelism
+//   - ParVectorRank is the nesting level on which to deploy vector parallelism
+//   - CurrentRank is the current nest level
+template <typename Rank, int ParThreadRank, int ParVectorRank, int CurrentRank>
+struct MDTeamRangeRanks {
+  using rank_type                            = int;
+  static constexpr Iterate iter              = Rank::outer_direction;
+  static constexpr rank_type total_rank      = Rank::rank;
+  static constexpr rank_type par_thread_rank = ParThreadRank;
+  static constexpr rank_type par_vector_rank = ParVectorRank;
+  static constexpr rank_type current_rank    = CurrentRank;
+
+  // We have to recursively process ranks [0..total_rank-1]
+  using RangeMode = MDTeamRangeMode<
+      (iter == Iterate::Right)
+          ? static_cast<MDTeamRangeEndRank>(current_rank == total_rank)
+          : static_cast<MDTeamRangeEndRank>(current_rank == -1),
+      static_cast<MDTeamRangeParThread>(current_rank == par_thread_rank),
+      static_cast<MDTeamRangeParVector>(current_rank == par_vector_rank)>;
+};
+
+// ParallelRank determines on which rank parallelization happens.
+//   - Rank is Kokkos::Rank<TotalRank, Iter>
+//     - TotalRank is the total number of loop nests
+//     - Iter is whether to go forward or backward through ranks (i.e. the
+//       iteration order for MDRangePolicy)
+//   - t_and_v determines whether both vector and thread parallelism is in use
+template <typename Rank, typename ExecSpace, MDTeamRangeThreadAndVector t_and_v>
+struct ParallelRank;
+
+#ifdef KOKKOS_ENABLE_SERIAL
+template <typename Rank, MDTeamRangeThreadAndVector t_and_v>
+struct ParallelRank<Rank, DefaultHostExecutionSpace, t_and_v> {
+  static constexpr bool isDirectionLeft =
+      (Rank::outer_direction == Iterate::Left);
+  static constexpr int par_rt  = isDirectionLeft ? Rank::rank - 1 : 0;
+  static constexpr int par_rv  = isDirectionLeft ? 0 : Rank::rank - 1;
+  static constexpr int invalid = -2;
+};
+#endif
+
+#ifdef KOKKOS_ENABLE_CUDA
+template <typename Rank, MDTeamRangeThreadAndVector t_and_v>
+struct ParallelRank<Rank, Cuda, t_and_v> {
+  static constexpr bool isDirectionLeft =
+      (Rank::outer_direction == Iterate::Left);
+  // If vector parallelism is in use deploy thread parallelism on
+  // the second to the last rank, otherwise, thread parallelism on the last rank
+  static constexpr int par_rt =
+      isDirectionLeft ? ((t_and_v) ? 1 : 0)
+                      : ((t_and_v) ? Rank::rank - 2 : Rank::rank - 1);
```

Comment:
These guys need to check that Rank>1 in order to allow for the 1 and Rank::rank-2 otherwise it needs to be 0, and Rank::rank-1

Reply:
ah I guess we just support rank 2 anyway for now ... maybe a follow up

Reply:
Yep, because we are now using Kokkos::Rank, it will assert that Ranks are range [2, 8].

https://github.com/kokkos/kokkos/blob/4f3ebbe219dd690c3938c56879758e0d55cfb8e2/core/src/Kokkos_Rank.hpp#L59-L72